pgx/src/spi.rs

```diff
@@ -378,8 +378,176 @@ impl SpiClient {
             current: -1,
         }
     }
+
+    /// sets up a cursor that will execute the specified query.
+    /// Rows may be then fetched using [`SpiCursor::fetch`]
+    ///
+    /// See [`SpiCursor`] docs for usage details.
+    pub fn open_cursor(
+        &mut self,
+        query: &str,
+        args: Option<Vec<(PgOid, Option<pg_sys::Datum>)>>,
+    ) -> SpiCursor {
+        let src = std::ffi::CString::new(query).expect("query contained a null byte");
+        let args = args.unwrap_or(vec![]);
+
+        let nargs = args.len();
+        let mut argtypes = vec![];
+        let mut datums = vec![];
+        let mut nulls = vec![];
+
+        for (argtype, datum) in args {
+            argtypes.push(argtype.value());
+
+            match datum {
+                Some(datum) => {
+                    // ' ' here means that the datum is not null
+                    datums.push(datum);
+                    nulls.push(' ' as std::os::raw::c_char);
+                }
+
+                None => {
+                    // 'n' here means that the datum is null
+                    datums.push(0);
+                    nulls.push('n' as std::os::raw::c_char);
+                }
+            }
+        }
+
+        let ptr = unsafe {
+            pg_sys::SPI_cursor_open_with_args(
+                std::ptr::null_mut(), // let postgres assign a name
+                src.as_ptr(),
+                nargs as i32,
+                argtypes.as_mut_ptr(),
+                datums.as_mut_ptr(),
+                nulls.as_ptr(),
+                false,
+                0,
+            )
+        };
+        SpiCursor { ptr }
+    }
+
+    /// find a cursor in transaction by name.
+    /// Name should usually be retrieved via [`SpiCursor::into_name`]
+    ///
+    /// See [`SpiCursor`] docs for usage details.
+    pub fn find_cursor(
+        &mut self,
+        name: String,
+    ) -> SpiCursor {
+        use pgx_pg_sys::AsPgCStr;
+
+        let ptr = unsafe {
+            pg_sys::SPI_cursor_find(name.as_pg_cstr())
+        };
+        if ptr.is_null() {
+            error!("cursor named \"{}\" not found", name);
+        }
+        SpiCursor { ptr }
+    }
+}
+
+/// Represents a Postgres cursor (internally, a portal), allowing to retrieve result rows a few
+/// at a time. Moreover, a cursor can be left open within a transaction, and accessed in
+/// multiple independent Spi sessions within the transaction.
+///
+/// A cursor can be created via [`SpiClient::open_cursor()`] from a query; a cursor left open (i.e.
+/// not [`Self::close()`]d) can be turned into a cursor name via [`Self::into_name()`] and later
+/// retrieved by name (in the same transaction) via [`SpiClient::find_cursor()`].
+///
+/// # Important notes about memory usage
+/// Result sets returned by [`SpiCursor::fetch()`] will not be freed until the current
+/// Spi session is complete; this is a Pgx limitation that might get lifted in the future.
```

Comment:
@zombodb I wonder if we could resolve this?

Reply:
I'm afraid I was way too terse, I could have clarified it better, let me try.

The issue I'm referring to is that zero-copy `FromDatum` implementations, such as `&str`, borrow the data from the `pg_sys::SPITupleTable` (wrapped in the `pgx::spi::SpiTupleTable`) returned by `select` and `fetch`, but this is not reflected by the type system (i.e. the returned `&str` has an inferred lifetime not bound by the `SpiTupleTable` it borrows from). 

As such, the borrow checker has no way to enforce proper lifetimes; this has been handled in the existing `pgx::spi` API by keeping all returned `SpiTupleTable`s around for the entire Spi session (by not exposing Postgres' `spi_freetuptable`) and having `Spi::connect` rely on `spi_finish` to release them all at the end, after copying any returned `FromDatum` in the upper memory context.

This works reasonably well for selects, where the user is already retrieving all the data at once.
Cursor users, on the other side, usually do so to avoid allocating memory for the entire result set all at once, so keeping `SpiTupleTable`s around might become a problem.

Thinking about it, I came up with a couple of options:

- recommending users to create separate Spi sessions if they care about freeing memory (current solution);
- having `fetch` take a closure as param, making it work more like `Spi::connect`, copying any returned data in the upper memory context and freeing the result set right away;
- having `fetch` return a different type, similar to `pgx::spi::PgTupleTable`, that yields borrowed data with a limited lifetime, forcing the user to copy it explicitly before dropping the source; this would have significant impact in term of dev experience I feel.

To start, I went with the simplest one, but I'm really interested in exploring it further based on your feedback.

Reply:
Makes sense 👍